cpp/ql/src/semmle/code/cpp/controlflow/StackVariableReachability.qll

```diff
@@ -3,7 +3,194 @@
  * reachability involving stack variables.
  */
 
-import cpp
+private import semmle.code.cpp.controlflow.Guards
+private import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+private newtype TCondition =
+  TTrue() or
+  TLogicalCondition(LogicalGuardCondition guard, boolean testIsTrue) { testIsTrue = [false, true] }
+
+/** A `GuardCondition` which appear in a control-flow path to a sink. */
+abstract private class LogicalGuardCondition extends GuardCondition {
+  LogicalGuardCondition() {
+    // Either the `GuardCondition` is part of the path from a source to a sink
+    revBbSuccessorEntryReaches0(_, this.getBasicBlock(), _, _, _)
+    or
+    // or it controls the basic block that contains the source node.
+    this.controls(any(BasicBlock bb | fwdBbEntryReachesLocally(bb, _, _, _)), _)
+  }
+
+  /**
+   * Holds if the truth of this logical expression having value `wholeIsTrue`
+   * implies that the truth of the child expression `part` has truth value `partIsTrue`.
+   */
+  abstract predicate impliesCondition(
+    LogicalGuardCondition e, boolean testIsTrue, boolean condIsTrue
+  );
+}
+
+private class BinaryLogicalGuardCondition extends LogicalGuardCondition, BinaryLogicalOperation {
+  override predicate impliesCondition(
+    LogicalGuardCondition e, boolean testIsTrue, boolean condIsTrue
+  ) {
+    this.impliesValue(e, testIsTrue, condIsTrue)
+  }
+}
+
+private class VariableGuardCondition extends LogicalGuardCondition, VariableAccess {
+  override predicate impliesCondition(
+    LogicalGuardCondition e, boolean testIsTrue, boolean condIsTrue
+  ) {
+    this = e and
```

Comment:
Is this case useful?

Reply:
Yes. If we instead implement this predicate as `none`, the conjunct in `LogicalCondition::impliesCondition`:
```codeql
this.getCondition().impliesCondition(globalValueNumber(other).getAnExpr(), /* ... */, /* ... */)
```
wouldn't hold when `this.getCondition()` is a `VariableGuardCondition`.